File: time_synchronization/python/ntp_sync.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (c) 2024 SICK AG, Waldkirch
#
# SPDX-License-Identifier: Unlicense

import argparse

from python_base.Control import Control
from shared.python.devices_config import get_device_config

from datetime import datetime, timezone

import time

def run_ntp_demo(ip_address: str, server_ip: str, ntp_port:int, timeout:int, cola_protocol: str, control_port: int):

    # tag::open_control_channel[]
    device_control = Control(ip_address, cola_protocol, control_port)
    device_control.open()
    # end::open_control_channel[]

    # tag::timestamp_before[]
    device_timestamp_ms = device_control.getDeviceTime()
    device_time = datetime.fromtimestamp(device_timestamp_ms/1000, tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')
    print(f"Device time before time synchronization: {device_time}")
    # end::timestamp_before[]

    # tag::wait_for_userinput[]
    # Ask user about NTP Master
    print("Running NTP Master is connected to the network? [Y/N]")
    user_input = input().strip().upper()
    # end::wait_for_userinput[]
    
    if user_input == 'Y':
        # tag::login[]
        # Login to the device for access rights to certain methods
        device_control.login(Control.USERLEVEL_SERVICE, 'CUST_SERV')
        # end::login[]
        
        # tag::set_ntp_server_ip[]
        device_control.setNtpClientServerAddress(server_ip)
        print(f"Set NTP server IP: {server_ip}")
        # end::set_ntp_server_ip[]

        # tag::set_ntp_server_port[]
        device_control.setNtpClientServerPort(ntp_port) # default is 123
        print(f"Set NTP port: {ntp_port}")
        # end::set_ntp_server_port[]

        # tag::set_ntp_server_timeout[]
        device_control.setNtpClientTimeout(timeout)
        print(f"Set NTP request timeout: {timeout}")
        # end::set_ntp_server_timeout[]

        # tag::enable_ntp[]
        device_control.setTimeSyncMode('NTP')
        print("Enabled NTP time synchronization")
        # end::enable_ntp[]
        device_control.logout()

        # tag::sleep[]
        time.sleep(5)
        print("Waiting for 5 seconds to be on the safe side.")
        # end::sleep[]

        # tag::timestamp_after[]
        device_timestamp_ms = device_control.getDeviceTime()
        device_time = datetime.fromtimestamp(device_timestamp_ms/1000, tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')
        print(f"Device time after time synchronization: {device_time}")
        # end::timestamp_after[]

    device_control.login(Control.USERLEVEL_SERVICE, 'CUST_SERV')
    # tag::disable_ntp[]
    device_control.setTimeSyncMode('NONE')
    print("Disabled NTP time synchronization")
    # end::disable_ntp[]

    # tag::logout_and_close[]
    device_control.logout()
    device_control.close()
    print("Logged out. Closed connection.")
    # end::logout_and_close[]

if __name__ == "__main__":

    parser = argparse.ArgumentParser(
        description="This script demonstrates how to synchronize the device time with NTP.")
    parser.add_argument('-i', '--ip_address', required=False, type=str,
                        default="192.168.1.10", help="The ip address of the device.")
    parser.add_argument('-s', '--server_ip', required=False, type=str,
                        default="127.0.0.1", help="The ip address of the ntp server.")
    parser.add_argument('-p', '--ntp_port', required=False, type=int,
                        default="123", help="The port of the ntp server.")
    parser.add_argument('-t', '--timeout', required=False, type=int,
                        default="10000", help="Timeout of the NTP client in microseconds.")
    args = parser.parse_args()

    cola_protocol, control_port, _ = get_device_config("Visionary-T Mini")

    run_ntp_demo(args.ip_address, args.server_ip, args.ntp_port, args.timeout, cola_protocol, control_port)

File: time_synchronization/cpp/ntp_sync.cpp
//
// Copyright (c) 2024 SICK AG, Waldkirch
//
// SPDX-License-Identifier: Unlicense

#include <algorithm>
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <string>

#include <iostream>
#include <sstream>
#include <thread>

#include <sick_visionary_cpp_base/CoLaParameterReader.h>
#include <sick_visionary_cpp_base/CoLaParameterWriter.h>
#include <sick_visionary_cpp_base/VisionaryControl.h>
#include <sick_visionary_cpp_base/VisionaryType.h>

#include "TimeStampHelper.h"
#include "exitcodes.h"

static ExitCode runNtpDemo(visionary::VisionaryType visionaryType,
                           const std::string&       ipAddress,
                           const std::string&       serverIp,
                           const uint16_t           serverPort)
{
  using namespace visionary;

  VisionaryControl visionaryControl(visionaryType);

  // Connect to devices control channel of the visionary device
  if (!visionaryControl.open(ipAddress))
  {
    std::fprintf(stderr, "Failed to open control connection to device\n");
    return ExitCode::eCommunicationError;
  }
  std::printf("Opened connection to Visionary device: %s\n", ipAddress.c_str());

  // Get device time before synchronization
  std::printf("Device time before time synchronization: ");
  printDeviceTime(visionaryControl);

  // Ask user about NTP Master
  std::string userInput;
  std::printf("\nWorking NTP Master is connected to the network? [Y/N]: ");
  std::getline(std::cin, userInput);

  // Convert to uppercase for consistency
  std::transform(userInput.begin(), userInput.end(), userInput.begin(), ::toupper);

  if (userInput == "Y" || userInput == "y")
  {
    // Login
    if (!visionaryControl.login(IAuthentication::UserLevel::AUTHORIZED_CLIENT, "CLIENT"))
    {
      std::fprintf(stderr, "Login failed\n");
      return ExitCode::eAuthenticationError;
    }

    // NTP settings
    CoLaCommand setNtpServerAddressCmnd = CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "ntpClientServerAddress")
                                            .parameterFlexString(serverIp)
                                            .build();
    CoLaCommand setNtpServerAddressCmndResponse = visionaryControl.sendCommand(setNtpServerAddressCmnd);
    if (setNtpServerAddressCmndResponse.getError() != CoLaError::OK)
    {
      std::fprintf(stderr, "Set NTP server address failed.\n");
      return ExitCode::eParamError;
    }
    else
    {
      std::printf("Set ntpClientServerAddress\n");
    }

    /// Set NTP server port
    CoLaCommand setNtpServerPortCmnd =
      CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "ntpClientServerPort").parameterUInt(serverPort).build();
    CoLaCommand setNtpServerPortCmndResponse = visionaryControl.sendCommand(setNtpServerPortCmnd);
    if (setNtpServerPortCmndResponse.getError() != CoLaError::OK)
    {
      std::fprintf(stderr, "Set NTP server port failed\n");
      return ExitCode::eParamError;
    }
    else
    {
      std::printf("Set ntpClientServerPort\n");
    }

    /// Enable NTP
    CoLaCommand enableNtpCommand = CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "timeSyncMode")
                                     .parameterUSInt(1)
                                     .build(); // 0 = NONE, 1 = NTP, 2 = PTP
    CoLaCommand enableNtpCommandResponse = visionaryControl.sendCommand(enableNtpCommand);
    if (enableNtpCommandResponse.getError() != CoLaError::OK)
    {
      std::fprintf(stderr, "Enable NTP time sync failed\n");
      return ExitCode::eParamError;
    }
    else
    {
      std::printf("Set timeSyncMode to NTP\n");
    }
    visionaryControl.logout();

    std::printf("Sleep for 5 seconds.\n");
    std::this_thread::sleep_for(std::chrono::milliseconds(5000));

    // Get device time after synchronization
    std::printf("Device time after time synchronization\n");
    printDeviceTime(visionaryControl);
  }
  
  /// Disable NTP
  if (!visionaryControl.login(IAuthentication::UserLevel::AUTHORIZED_CLIENT, "CLIENT"))
  {
    std::fprintf(stderr, "Login failed\n");
    return ExitCode::eAuthenticationError;
  }
  CoLaCommand disableNtpCommand =
    CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "timeSyncMode").parameterUSInt(0).build();
  CoLaCommand disableNtpCommandResponse = visionaryControl.sendCommand(disableNtpCommand);
  if (disableNtpCommandResponse.getError() != CoLaError::OK)
  {
    std::fprintf(stderr,"Disable NTP time sync failed\n");
    return ExitCode::eParamError;
  }
  else
  {
    std::printf("Disabled NTP timeSyncMode\n");
  }
  
  visionaryControl.logout();
  visionaryControl.close();
  std::printf("Logout and closed connection.\n");
  return ExitCode::eOk;
}

int main(int argc, char* argv[])
{
  std::string   deviceIpAddr("192.168.1.10");
  std::string   serverIpAddr("192.168.136.100");
  std::uint16_t serverPort{123};

  visionary::VisionaryType visionaryType(visionary::VisionaryType::eVisionaryTMini);

  bool showHelpAndExit = false;

  ExitCode exitCode = ExitCode::eOk;

  for (int i = 1; i < argc; ++i)
  {
    std::istringstream argstream(argv[i]);

    if (argstream.get() != '-')
    {
      showHelpAndExit = true;
      exitCode        = ExitCode::eParamError;
      break;
    }
    switch (argstream.get())
    {
      case 'h':
        showHelpAndExit = true;
        break;
      case 'i':
        argstream >> deviceIpAddr;
        break;
      case 's':
        argstream >> serverIpAddr;
        break;
      case 'p':
        argstream >> serverPort;
        break;
      default:
        showHelpAndExit = true;
        break;
    }
  }

  if (showHelpAndExit)
  {
    std::cout << argv[0] << " [option]*"
              << "\n";
    std::cout << "where option is one of"
              << "\n";
    std::cout << "-h          show this help and exit"
              << "\n";
    std::cout << "-i<IP>      connect to the device with IP address <IP>; default is " << deviceIpAddr << "\n";
    std::cout << "-s<sIP>     connect to the NTP server with IP address <sIP>; default is " << serverIpAddr << "\n";
    std::cout << "-p<Prt>     connect to the NTP server with port <Prt>; default is " << serverPort << "\n";

    return static_cast<int>(exitCode);
  }

  exitCode = runNtpDemo(visionaryType, deviceIpAddr, serverIpAddr, serverPort);

  std::cout << "exit code " << static_cast<int>(exitCode) << "\n";

  return static_cast<int>(exitCode);
}
